src/providers/hana/qgshanafeatureiterator.cpp

```diff
@@ -0,0 +1,437 @@
+    {
+      QgsHanaExpressionCompiler compiler = QgsHanaExpressionCompiler( mSource );
+      QgsExpression expression = clause.expression();
+      if ( compiler.compile( &expression ) == QgsSqlExpressionCompiler::Complete )
+      {
+        QString part;
+        part = compiler.result();
+        part += clause.ascending() ? QStringLiteral( " ASC" ) : QStringLiteral( " DESC" );
+        part += clause.nullsFirst() ? QStringLiteral( " NULLS FIRST" ) : QStringLiteral( " NULLS LAST" );
+        orderByParts << part;
+      }
+      else
+      {
+        // Bail out on first non-complete compilation.
+        // Most important clauses at the beginning of the list
+        // will still be sent and used to pre-sort so the local
+        // CPU can use its cycles for fine-tuning.
+        mOrderByCompiled = false;
+        break;
+      }
+    }
+  }
+  else
+  {
+    mOrderByCompiled = false;
+  }
+#endif
+
+  if ( !mOrderByCompiled )
+    limitAtProvider = false;
+
+  bool subsetOfAttributes = mRequest.flags() & QgsFeatureRequest::SubsetOfAttributes;
+  QgsAttributeList attrs = ( subsetOfAttributes ) ?
+                           request.subsetOfAttributes() : mSource->mFields.allAttributesList();
+
+  if ( subsetOfAttributes )
+  {
+    // Ensure that all attributes required for expression filter are fetched
+    if ( mRequest.filterType() == QgsFeatureRequest::FilterExpression )
+    {
+      //ensure that all fields required for filter expressions are prepared
+      QSet<int> attributeIndexes = request.filterExpression()->referencedAttributeIndexes( mSource->mFields );
+      attributeIndexes += attrs.toSet();
+      attrs = attributeIndexes.toList();
+    }
+
+    if ( !mRequest.orderBy().isEmpty() )
+    {
+      // Ensure that all attributes required for order by are fetched
+      const auto usedAttributeIndices = mRequest.orderBy().usedAttributeIndices( mSource->mFields );
+      for ( int attrIndex : usedAttributeIndices )
+      {
+        if ( !attrs.contains( attrIndex ) )
+          attrs << attrIndex;
+      }
+    }
+  }
+
+  Q_FOREACH ( int i, attrs )
+  {
+    QString fieldname = mSource->mFields.at( i ).name();
+    if ( mFidColumn == fieldname )
+      continue;
+    mAttributesToFetch.append( i );
+  }
+
+  QString sqlFields = "";
+
+  // Add feature id column
+  if ( !mFidColumn.isEmpty() )
+    sqlFields += QgsHanaUtils::quotedIdentifier( mFidColumn ) + ",";
+
+  Q_FOREACH ( int i, mAttributesToFetch )
+  {
+    QString fieldname = mSource->mFields.at( i ).name();
+    sqlFields += QStringLiteral( "%1," ).arg( QgsHanaUtils::quotedIdentifier( fieldname ) );
+  }
+  mHasAttributes = !mAttributesToFetch.isEmpty();
+
+  // Add geometry column
+  if ( ( !( request.flags() & QgsFeatureRequest::NoGeometry )
+         || ( request.filterType() == QgsFeatureRequest::FilterExpression &&
+              request.filterExpression()->needsGeometry() ) ) && mSource->isSpatial() )
```

Comment:
This condition is rather hard to understand. Can you split it up by evaluating some expressive booleans before, e.g.
```c++
const bool geometryRequested = (request.flags() & QgsFeatureRequest::NoGeometry) != 0;
//...
```

Reply:
Done in 38bdec5.